src/java.base/macosx/classes/sun/nio/fs/MacOSXWatchService.java

```diff
@@ -0,0 +1,634 @@
+    MacOSXWatchService() throws IOException {
+        runLoopThread = new CFRunLoopThread();
+        Thread t = InnocuousThread.newThread("FileSystemWatcher", runLoopThread);
+        t.setDaemon(true);
+        t.start();
+
+        // In order to be able to schedule any FSEventStream's, a reference to a run loop is required.
+        runLoopThread.waitForRunLoopRef();
+    }
+
+    @SuppressWarnings("removal")
+    @Override
+    WatchKey register(Path dir,
+                      WatchEvent.Kind<?>[] events,
+                      WatchEvent.Modifier... modifiers) throws IOException {
+        checkIsOpen();
+
+        final UnixPath unixDir = (UnixPath)dir;
+        final Object dirKey    = checkPath(unixDir);
+        final EnumSet<FSEventKind>   eventSet    = FSEventKind.setOf(events);
+        final EnumSet<WatchModifier> modifierSet = WatchModifier.setOf(modifiers);
+
+        try {
+            return AccessController.doPrivileged(
+                    new PrivilegedExceptionAction<MacOSXWatchKey>() {
+                @Override
+                public MacOSXWatchKey run() throws IOException {
+                    return doPrivilegedRegister(unixDir, dirKey, eventSet, modifierSet);
+                }
+            });
+        } catch (PrivilegedActionException pae) {
+            final Throwable cause = pae.getCause();
+            if (cause instanceof IOException ioe)
+                throw ioe;
+            throw new AssertionError(pae);
+        }
+    }
+
+    private MacOSXWatchKey doPrivilegedRegister(UnixPath unixDir,
+                                                Object dirKey,
+                                                EnumSet<FSEventKind> eventSet,
+                                                EnumSet<WatchModifier> modifierSet) throws IOException {
+        synchronized (closeLock()) {
+            checkIsOpen();
+
+            synchronized (watchKeysLock) {
+                MacOSXWatchKey watchKey = dirKeyToWatchKey.get(dirKey);
+                final boolean keyForDirAlreadyExists = (watchKey != null);
+                if (keyForDirAlreadyExists) {
+                    eventStreamToWatchKey.remove(watchKey.getEventStreamRef());
+                    watchKey.disable();
+                } else {
+                    watchKey = new MacOSXWatchKey(this, unixDir, dirKey);
+                    dirKeyToWatchKey.put(dirKey, watchKey);
+                }
+
+                watchKey.enable(runLoopThread, eventSet, modifierSet);
+                eventStreamToWatchKey.put(watchKey.getEventStreamRef(), watchKey);
+                watchKeysLock.notify(); // So that run loop gets running again
+                                        // if stopped due to lack of event streams
+                return watchKey;
+            }
+        }
+    }
+
+    /**
+     * Invoked on the CFRunLoopThread by the native code to report directories
+     * that need to be re-scanned.
+     */
+    private void callback(final long eventStreamRef,
```

Comment:
Could this be named something more descriptive than `callback`, perhaps even `handleEvents` even though there is a method of the same name defined in `MacOSXWatchKey`?

Reply:
Sure.